aten/src/ATen/native/cuda/PersistentSoftmax.cuh

```diff
@@ -0,0 +1,385 @@
+    batches_per_warp = (next_power_of_two <= 128) ? 2 : 1;
+
+    switch (log2_elements) {
+    case 0: // 1
+        kernel = &softmax_warp_forward<input_t, output_t, acc_t, 2,1,1, is_log_softmax>;
+        break;
+    case 1: // 2
+        kernel = &softmax_warp_forward<input_t, output_t, acc_t, 2,1,2, is_log_softmax>;
+        break;
+    case 2: // 4
+        kernel = &softmax_warp_forward<input_t, output_t, acc_t, 2,1,4, is_log_softmax>;
+        break;
+    case 3: // 8
+        kernel = &softmax_warp_forward<input_t, output_t, acc_t, 2,1,8, is_log_softmax>;
+        break;
+    case 4: // 16
+        kernel = &softmax_warp_forward<input_t, output_t, acc_t, 2,1,16, is_log_softmax>;
+        break;
+    case 5: // 32
+        kernel = &softmax_warp_forward<input_t, output_t, acc_t, 2,1,32, is_log_softmax>;
+        break;
+    case 6: // 64
+        kernel = &softmax_warp_forward<input_t, output_t, acc_t, 2,2,32, is_log_softmax>;
+        break;
+    case 7: // 128
+        kernel = &softmax_warp_forward<input_t, output_t, acc_t, 2,4,32, is_log_softmax>;
+        break;
+    case 8: // 256
+        kernel = &softmax_warp_forward<input_t, output_t, acc_t, 1,8,32, is_log_softmax>;
+        break;
+    case 9: // 512
+        kernel = &softmax_warp_forward<input_t, output_t, acc_t, 1,16,32, is_log_softmax>;
+        break;
+    case 10: // 1024
+        kernel = &softmax_warp_forward<input_t, output_t, acc_t, 1,32,32, is_log_softmax>;
+        break;
+    default:
+        return false;
+    }
+    return true;
+}
+
+template<typename input_t, typename output_t, typename acc_t, bool is_log_softmax>
+bool dispatch_softmax(output_t *dst, const input_t *src, int softmax_elements, int softmax_elements_stride, int batch_count)
+{
+    if (softmax_elements == 0) {
+        return true;
+    } else if (softmax_elements <= 1024) {
+        // compute function index. there's a function for each power of two size up to 1024.
+        int log2_elements = 0;
+        while ((1 << log2_elements) < softmax_elements) ++log2_elements;
+
+        softmax_forward_func<input_t, output_t> kernel;
+        int warp_size, batches_per_warp;
+        if (!warp_softmax_kernel<input_t, output_t, acc_t, is_log_softmax>(log2_elements, warp_size, batches_per_warp, kernel)) {
+            return false;
+        }
+
+        // use 128 threads per block to maximimize gpu utilization
+        constexpr int threads_per_block = 128;
+
+        // compute warps per block.
+        int warps_per_block = (threads_per_block / warp_size);
+
+        // compute launch size
+        int batches_per_block = warps_per_block * batches_per_warp;
+        int blocks = (batch_count + batches_per_block - 1) / batches_per_block;
+        dim3 threads(warp_size, warps_per_block, 1);
+
+        // launch
+        kernel<<<blocks, threads>>>(dst, src, batch_count, softmax_elements_stride, softmax_elements);
+        return true;
+    }
+    return false;
```

Comment:
might make sense to do TORCH_INTERNAL_ASSERT on softmax_elements <=1024.

Reply:
That's actually a better idea than asserting on the return value from dispatch_softmax. Will do.